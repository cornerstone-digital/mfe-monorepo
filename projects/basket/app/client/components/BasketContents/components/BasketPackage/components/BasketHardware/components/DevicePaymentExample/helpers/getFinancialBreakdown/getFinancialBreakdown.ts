import formatAmount from '@helpers/formatAmount'
import formatCurrency from '@helpers/formatCurrency'
import { DeviceCosts } from '../../DevicePaymentExample.types'

const getFinancialBreakdown = (
  devicePaymentPlan: BasketV2.DevicePaymentPlan,
  airtimePrice: string = '0',
  priceKey: keyof BasketV2.DevicePrice,
) => {
  const deviceCosts: DeviceCosts = {
    upFrontPrice: formatCurrency(devicePaymentPlan?.upFrontPrice?.[priceKey] || '0'),
    monthlyPrice: formatCurrency(devicePaymentPlan.monthlyPrice?.[priceKey] || '0'),
    totalHandsetPrice: formatCurrency(devicePaymentPlan.totalHandsetPrice?.[priceKey] || '0'),
    totalHandsetCredit: formatCurrency(devicePaymentPlan.totalHandsetCredit?.[priceKey] || '0'),
    interestRate: `${formatAmount(devicePaymentPlan.interestRate || 0)}%`,
    aprRepresentative: `${formatAmount(devicePaymentPlan.aprRepresentative || '0')}%`,
    creditor: 'Vodafone Ltd.',
    duration: devicePaymentPlan.duration?.value || '0',
  }

  const airtimeCosts = {
    monthlyCost: formatCurrency(airtimePrice || '0'),
  }

  const monthlyAirtime = parseFloat(airtimePrice || '0')
  const monthlyHardware = parseFloat(devicePaymentPlan?.monthlyPrice?.[priceKey] || '0')
  const monthlyTotal = (monthlyAirtime + monthlyHardware).toFixed(2)

  return {
    deviceCosts,
    airtimeCosts,
    total: {
      duration: devicePaymentPlan?.duration?.value,
      upfrontCost: formatCurrency(devicePaymentPlan?.upFrontPrice?.[priceKey] || '0'),
      monthlyTotal: formatCurrency(monthlyTotal),
    },
  }
}

export default getFinancialBreakdown
