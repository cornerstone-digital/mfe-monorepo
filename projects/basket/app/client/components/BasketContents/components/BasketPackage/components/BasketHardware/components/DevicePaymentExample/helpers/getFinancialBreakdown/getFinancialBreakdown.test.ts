import getFinancialBreakdown from './getFinancialBreakdown'
import mockPaymentPlan from '../../mocks/paymentPlan.mock.json'

describe('getFinancialBreakdown', () => {
  it.each`
    airtimePrice | priceKey
    ${undefined} | ${'net'}
    ${'14.99'}   | ${'net'}
    ${undefined} | ${'gross'}
    ${'14.99'}   | ${'gross'}
  `('should correctly format data with $priceKey prices with $airtimePrice airtime price', ({ airtimePrice, priceKey }) => {
    expect(getFinancialBreakdown(mockPaymentPlan, airtimePrice, priceKey)).toMatchSnapshot()
  })
})
