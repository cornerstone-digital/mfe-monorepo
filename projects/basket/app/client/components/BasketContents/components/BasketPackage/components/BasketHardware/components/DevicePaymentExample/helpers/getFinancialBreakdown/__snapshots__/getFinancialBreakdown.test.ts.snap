// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getFinancialBreakdown should correctly format data with gross prices with 14.99 airtime price 1`] = `
Object {
  "airtimeCosts": Object {
    "monthlyCost": "£14.99",
  },
  "deviceCosts": Object {
    "aprRepresentative": "0%",
    "creditor": "Vodafone Ltd.",
    "duration": "7",
    "interestRate": "0%",
    "monthlyPrice": "£166.71",
    "totalHandsetCredit": "£1167",
    "totalHandsetPrice": "£1320",
    "upFrontPrice": "£153",
  },
  "total": Object {
    "duration": "7",
    "monthlyTotal": "£181.70",
    "upfrontCost": "£153",
  },
}
`;

exports[`getFinancialBreakdown should correctly format data with gross prices with undefined airtime price 1`] = `
Object {
  "airtimeCosts": Object {
    "monthlyCost": "£0",
  },
  "deviceCosts": Object {
    "aprRepresentative": "0%",
    "creditor": "Vodafone Ltd.",
    "duration": "7",
    "interestRate": "0%",
    "monthlyPrice": "£166.71",
    "totalHandsetCredit": "£1167",
    "totalHandsetPrice": "£1320",
    "upFrontPrice": "£153",
  },
  "total": Object {
    "duration": "7",
    "monthlyTotal": "£166.71",
    "upfrontCost": "£153",
  },
}
`;

exports[`getFinancialBreakdown should correctly format data with net prices with 14.99 airtime price 1`] = `
Object {
  "airtimeCosts": Object {
    "monthlyCost": "£14.99",
  },
  "deviceCosts": Object {
    "aprRepresentative": "0%",
    "creditor": "Vodafone Ltd.",
    "duration": "7",
    "interestRate": "0%",
    "monthlyPrice": "£138.93",
    "totalHandsetCredit": "£972.50",
    "totalHandsetPrice": "£1100",
    "upFrontPrice": "£127.50",
  },
  "total": Object {
    "duration": "7",
    "monthlyTotal": "£153.92",
    "upfrontCost": "£127.50",
  },
}
`;

exports[`getFinancialBreakdown should correctly format data with net prices with undefined airtime price 1`] = `
Object {
  "airtimeCosts": Object {
    "monthlyCost": "£0",
  },
  "deviceCosts": Object {
    "aprRepresentative": "0%",
    "creditor": "Vodafone Ltd.",
    "duration": "7",
    "interestRate": "0%",
    "monthlyPrice": "£138.93",
    "totalHandsetCredit": "£972.50",
    "totalHandsetPrice": "£1100",
    "upFrontPrice": "£127.50",
  },
  "total": Object {
    "duration": "7",
    "monthlyTotal": "£138.93",
    "upfrontCost": "£127.50",
  },
}
`;
